File: 2563-count-the-number-of-fair-pairs/two-pointers/python/solution.py
class Solution:
    def countFairPairs(self, nums: List[int], lower: int, upper: int) -> int:
        nums.sort()

        # Helper to count pairs with sum less than or equal to upper_bound
        def count_pairs(upper_bound) -> int: 
            count = 0
            left, right = 0, len(nums) - 1

            while left < right:
                pair_sum = nums[left] + nums[right]
                if pair_sum <= upper_bound:
                    count += right - left
                    left += 1
                else:
                    right -= 1
            return count

        return count_pairs(upper) - count_pairs(lower - 1)

File: 2563-count-the-number-of-fair-pairs/two-pointers/java/solution.java
class Solution {
    public long countFairPairs(int[] nums, int lower, int upper) {
        Arrays.sort(nums);
        return countPairs(nums, upper) - countPairs(nums, lower - 1);
    }

    // Helper to count pairs with sum less than or equal to upperBound
    private long countPairs(int[] nums, int upperBound) {
        long count = 0;
        int left = 0;
        int right = nums.length - 1;

        while (left < right) {
            int pairSum = nums[left] + nums[right];
            if (pairSum <= upperBound) {
                count += (right - left);
                left++;
            } else {
                right--;
            }
        }
        return count;
    }
    
}
